src/util/schnorr.rs

```diff
@@ -85,4 +89,75 @@ impl TweakedPublicKey {
         &self.0
     }
 
-}
\ No newline at end of file
+}
+
+/// A BIP340-341 serialized schnorr signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SchnorrSig {
+    /// The underlying schnorr signature
+    pub sig: secp256k1::schnorrsig::Signature,
+    /// The corresponding hash type
+    pub hash_ty: SchnorrSigHashType,
+}
+
+impl SchnorrSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, SchnorrSigError> {
+        if sl.len() == SCHNORRSIG_SIGNATURE_SIZE {
+            // default type
+            let sig = secp256k1::schnorrsig::Signature::from_slice(sl)?;
+            return Ok( SchnorrSig { sig, hash_ty : SchnorrSigHashType::Default });
+        }
+        let (hash_ty, sig) = sl.split_last()
+            .ok_or_else(|| secp256k1::Error::InvalidSignature)?;
+        let hash_ty = SchnorrSigHashType::from_u8(*hash_ty)
+            .map_err(|_| SchnorrSigError::InvalidSighashType(*hash_ty))?;
+        let sig = secp256k1::schnorrsig::Signature::from_slice(sig)?;
+        Ok(SchnorrSig { sig, hash_ty })
```

Comment:
I am not conveinced that case when we have 36 byte signature should error with `InvalidSighashType` in random manner :(

I mean that error that will fire in fact will depend on the value of the last byte even for signatures of incorrect length which certainly do not have sighash attached.

We either need dedicated error (Signature and sighash are incorrect) - or should distinguish incorrect sighash in 65 byte sig from incorrect sig in <64 case.

Reply:
How do you feel about the new changes?

Reply:
That's exactly what I hoped for :)